Storage API put-get

1devNdogs opened the pull request:
# Summary

From issue #2 

- **Identity**:  
  - `fromWeb3`: Creates an **Identity** instance from `web3Accounts`.
  - `fromUri`: Creates an **Identity** from a secret URI.

- **SubspaceClient**: 
  - `connect`: it create providers to call node and farmer methods.
  - `putObject`: from an object, call put and return objectId
  - `getObject`: from an objectId return the object

- Readme with project description and instructions.

Next PR contains examples for front and backend projects.

nazar-pc commented on src/constants/index.ts:
```diff
@@ -0,0 +1,5 @@
+const NO_WEB3 = "Web3 not enabled or available.";
+const NO_ACCOUNT = "No accounts available";
+const CRYPTO_LOAD_FAIL = "Crypto wait ready failed";
```
Let's use proper errors derived from `Error` class instead of string constants. Otherwise we will not have backtraces and other good things.

1devNdogs replied:
I made this update.
```
const NO_WEB3 = new Error("Web3 not enabled or available.");
const NO_ACCOUNT = new Error("No accounts available");
const CRYPTO_LOAD_FAIL = new Error("Crypto wait ready failed");
```
I created these for some specific scenarios that do not throw errors, but I wonder if this is the correct usage, for example, if web3 is enabled but the user does not have accounts available:

  ```
 const allWeb3Accounts = await web3Accounts(accountSettings);
 if (!allWeb3Accounts || allWeb3Accounts.length === 0) throw NO_ACCOUNT;

```

nazar-pc replied:
No, you want to create an instance of an error where you throw it since the backtrace will point to the line where instance was created, not where it was thrown from.

1devNdogs replied:
[#de61605ed2dd246a99ef167984ec4317292d661f](https://github.com/subspace/subspace.js/pull/3/commits/de61605ed2dd246a99ef167984ec4317292d661f)

nazar-pc commented on src/identity.ts:
```diff
@@ -0,0 +1,146 @@
+   */
+  public static fromWeb3 = async (
+    address?: string | Uint8Array
+  ): Promise<Identity> => {
```
This signature and function are a bit confusing. It is both async and arrow function and returns manually constructed promise.

Let's do this instead:
```typescript
  public static async fromWeb3(address?: string | Uint8Array): Promise<Identity> {
```

And refactor body to not construct promise manually as I believe it is not required here.

nazar-pc replied:
Same comment applies to all other methods.

1devNdogs replied:
I also removed the `fromKeypair` method as you suggested in the overall comment.

nazar-pc commented on src/identity.ts:
```diff
@@ -0,0 +1,146 @@
+   * @param keyringPair  KeyringPair object.
+   * @return Identity
+   */
+  public constructor(
```
Should this be public though? We have 3 constructors above, I believe this should be private to not confuse people.

nazar-pc commented on src/identity.ts:
```diff
@@ -0,0 +1,146 @@
+    });
+  };
+
+  /**
+   * @name fromUri
+   * @summary Creates an Identity from a secret uri.
+   * @description SURI format for specifying secret keys.
+   * <secret>/<soft-key>//<hard-key>///<password>`
+   * `/<soft-key>` and `//<hard-key>` maybe repeated and mixed.
+   * `///password` may be omitted.
+   * The secret can be a hex string, mnemonic phrase or a string (to be padded).
+   * @param suri Example: `//Alice///password` or a mnemonic phrase.
+   * @return Promise<Identity>
+   */
```
Does this render properly in IDE?

1devNdogs replied:
It does not render 100% properly, I have updated it to be properly rendered.

nazar-pc commented on package.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "subspace.js",
+  "version": "1.0.0",
```
Really early for 1.0 :slightly_smiling_face: 
```suggestion
  "version": "0.1.0",
```

nazar-pc commented on src/subspaceClient.ts:
```diff
@@ -0,0 +1,130 @@
+    wsRpcEndpoint?: string,
+    wsRpcEndpointFarmer?: string
+  ): Promise<SubspaceClient> {
+    return new Promise<SubspaceClient>(async (resolve, reject) => {
```
The function is already async, no need to wrap into another function and in explicit promise on top of that

nazar-pc replied:
And the same for other methods below.

1devNdogs replied:
I have applied this to the connect method.
But for putObject and getObject I have removed the async and kept the Promise.

nazar-pc commented on src/subspaceClient.ts:
```diff
@@ -0,0 +1,130 @@
+   * @param signer
+   * @return boolean | any
+   */
+  public setSigner(signer: Signer): boolean | any {
```
Let's not return `any`. If it is an error - throw it, I don't see how it is useful to return `boolean | any`. Should just return `void` and throw an error just like it does by default.

1devNdogs replied:
Updated.
```  
public setSigner(signer: Signer): void {
    return this.api.setSigner(signer);
}
```

nazar-pc commented on src/subspaceClient.ts:
```diff
@@ -0,0 +1,130 @@
+
+        resolve(new SubspaceClient(api, farmerProvider, identity));
+      } catch (e) {
+        reject(e.message);
```
This loses the whole backtrace and makes debugging experience pretty bad, please don't do this.

nazar-pc commented on src/subspaceClient.ts:
```diff
@@ -0,0 +1,130 @@
+  /**
+   * @name putObject
+   * @summary Using a keypair from Identity Object, submit an extrinsic using objectStore module.
+   * @param object an 8 bytes array that contains the data to store.
```
Hm, it is not 8 bytes array, I guess you meant `Uint8Array`?

1devNdogs replied:
Yes, also fixed on Readme. Thanks!

nazar-pc commented on tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "lib": ["ES2021", "dom"],
+    "rootDir": "src"
+  },
+  "include": ["./src/**/*"]
```
:slightly_smiling_face: 

```suggestion
  "include": ["src"]
```

BTW, why not copy tsconfig.json from relayer where strict mode is already configured and other things?

1devNdogs replied:
I have updated from relayer and added  `"declaration": true,` also included your change.

nazar-pc commented on README.md:
```diff
@@ -1 +1,101 @@
+
+Install dependencies
+
+- `npm install`
```
Always `npm ci`

nazar-pc reviewed (commented):
Overall:
* this is a good start, with minor tweaks we can publish 0.1.0
* I have concerns about JsDoc syntax, I'm not sure it is valid and things like `@name` are honestly fully redundant, but we'll see when we add some tooling for auto-generating API documentation
* readme talks about a lot of things that are honestly irrelevant as user of this library doesn't interact with node, farmer or Web3 libraries directly (I also think we might want to remove `fromKeypair` constructor for identity for now to simplify it)

1devNdogs commented:
> Overall:
> 
> * this is a good start, with minor tweaks we can publish 0.1.0
> * I have concerns about JsDoc syntax, I'm not sure it is valid and things like `@name` are honestly fully redundant, but we'll see when we add some tooling for auto-generating API documentation
> * readme talks about a lot of things that are honestly irrelevant as user of this library doesn't interact with node, farmer or Web3 libraries directly (I also think we might want to remove fromKeypair` constructor for identity for now to simplify it)

- Regarding the jsDocs syntax, I added @name and details for future uses with some autogeneration tools, trying to follow the same syntax as Polkadot js methods. We can totally improve them.

- Regarding the root README.md, I removed irrelevant descriptions, trying to include only library classes and methods.

- Also removed `fromKeypair` method on the Identity class and docs.

I am validating other changes and will push them to cover your feedback.

isSerge commented on src/subspaceClient.ts:
```diff
@@ -0,0 +1,114 @@
+        this.setSigner(injected.signer);
+      }
+
+      this.api.tx.objectStore
```
probably we need to unsubscribe here

1devNdogs replied:
Yeah, I was checking the RPC calls, and doing this it calls `author_unwatchExtrinsic` after resolve and reject, Thanks!

```typescript
 const unsubscribe = await this.api.tx.objectStore
    .put(u8aToHex(object))
    .signAndSend(account, ({ status, events, isError }) => {
         ....
            resolve(event.data[1].toString().slice(2));
            unsubscribe();
         ....   
      } else if (isError) {
        reject(new Error(`isError: ${isError}`));
        unsubscribe();
      }
    });
```

nazar-pc commented on src/subspaceClient.ts:
```diff
@@ -0,0 +1,114 @@
+    wsRpcEndpoint?: string,
+    wsRpcEndpointFarmer?: string
+  ): Promise<SubspaceClient> {
+    const farmerProvider = new WsProvider(
```
I just noticed you use `WsProvider` form Polkadot.js for connection to farmer.
Honestly I'm surprised it works, but since it does, why not I guess :slightly_smiling_face:

1devNdogs replied:
I was using another [lib](https://github.com/elpheria/rpc-websockets ) to connect that also worked. But after a while, I realized that `WsProvider` had this `send` method. 
It saves us a dependency

nazar-pc reviewed (approved):
Looks much better, thanks!
